command line interface

tgross opened the pull request:
The Nomad team has a fork of this project (https://github.com/shoenig/hc-install) with the command line interface implemented (and Vault added). We use this to install Consul and Vault as part of CI (ex [github/workflows/test-core.yaml#L148-L149](https://github.com/hashicorp/nomad/blob/894c2e61dd03c9fad460d64a130dab807c4e428b/.github/workflows/test-core.yaml#L148-L149)). Rather than having to keep up with release API changes, we'd like to upstream our work.

Fixes https://github.com/hashicorp/hc-install/issues/13

radeksimko reviewed (commented):
Thanks for the PR @tgross 

I took the liberty and cherry-picked your commit adding Vault to `main` and added a test for the build (which tends to be most fragile of installation methods) https://github.com/hashicorp/hc-install/commit/bbc7232afe29a0b458f392405987c4a015ba7b54

Regarding the CLI, I have some thoughts on the UX - will review this 🔜 .

radeksimko commented:
The library started out with a relatively flexible (Go) API and a minimal dependency tree, both of which is also what motivated https://github.com/hashicorp/hc-install#hc-install-is-not-a-package-manager and make it possible for e.g. Terraform SDK to use this without much cost. It would be great if we could keep it that way.

I'm happy with `hc-install` providing some minimal CLI interface, but I wouldn't want it to step into the area of https://github.com/gulducat/hashi-bin I think if we stick with the mentioned design guidelines, it should be possible and relatively easy to build tools such as `hashi-bin`/tfenv and any other similar ones top of this library by leveraging the Go API.

With all that in mind:

1. How would you feel about the UX being `hc-install install <product> [-version=<VERSION>]`?
   - This leaves room for `-version-constraint` or `-build-rev` should we need to implement it in the future.
   - This also leaves room for e.g. `hc-install list-products` should we need to implement it in the future.
   - We could treat `hc-install install <product>` as "install latest version of <product>". I'm not suggesting do now, but it just seems more natural to make a named flag optional than a positional argument.
2. How about we create the `product.Product{}` on-the-fly? Since we're limited to installing just from releases.hashicorp.com, we can technically support almost any product from there (any which is packaged as ZIP), as installation from there only requires `Name` and `BinaryName`, we don't need `GetVersion()` nor `BuildInstructions` https://github.com/hashicorp/hc-install/blob/2695df3f02e3daf30380aad43fcaa54ad6271855/product/product.go#L10-L24
3. Would you mind making the installation path a flag? e.g. `-path`?
   - You can still pass `-path=$GOBIN` or `-path=$GOPATH/bin` downstream if you like.
   - This avoids setting any expectations on where the binary gets installed and avoids excuse for future PRs adding various other locations and us having to decide preferences between any of these and address other kinds of related problems, like where/whether does it need to be tracked and whether the CLI should also have uninstall or upgrade command. Generally avoids the scope-creep into a package manager territory.
   - Should this be a required flag? If not, what's the default?🤔 We could install by default into [`os.TempDir()`](https://pkg.go.dev/os#TempDir), just like the Go API does, the only question is how do we communicate that binary path back.
     - We could make it the only output on STDOUT and send other log lines with progress etc. to STDERR?
     - We could just print it out along with some other human-friendly output and introduce `-json` flag for JSON output, so downstream can pipe it into `jq`?

--- 

I understand I'm suggesting some significant changes to your PR, but I hope the context and explanation above can help understand why.

Either way - thank you for kicking this off and raising the PR and thank you for the patience. Let me know what you think!

shoenig commented:
This all sounds reasonable @radeksimko - but I am wondering about the default value for `-path`. 

> We could install by default into [os.TempDir()](https://pkg.go.dev/os#TempDir), just like the Go API does

What is this referring to? I was basing the idea off of `go install` which installs into `GOBIN` or `GOPATH/bin`.

kmoe commented:
Thanks for this! 

Agree with @radeksimko's comments above. In particular regarding the default installation path, note that as per the README,
>hc-install does not:
>Determine suitable installation path based on target system. e.g. in /usr/bin or /usr/local/bin on Unix based system.

I think therefore the default path for downloaded binaries should be the current directory, the analogy being `curl`/`wget` rather than `go install` or `apt get`.

`os.TempDir()` (as in [`LatestVersion.Install()`](https://github.com/hashicorp/hc-install/blob/af09eeeab48c4163071c2672264e43c2ae22eeb1/releases/latest_version.go#L82) for example) is also a good option, but it is less idiomatic for a CLI tool to output the path of a downloaded file.

radeksimko commented:
@kmoe I like that idea 👍🏻  and I like your analogy with `curl`/`wget` - that also avoids the need for communicating an unpredictable random temp dir path back. All the consumer needs to know is name of the binary, which it presumably already does, so we can avoid the complexity of a predictable strict output format.

@tgross How does that sound as a default option for `-path`?

tgross commented:
> @tgross How does that sound as a default option for -path?

Works for me. I'll take a pass at updating this PR with your suggestions sometime in the next couple days probably.

tgross commented:
I've updated this PR as follows:
* rebased on main to make it just the CLI changes.
* `hc-install install` is now a subcommand.
* The subcommand takes a required `-version` flag.
* The subcommand takes an optional `-flag` path, defaulting to installing in the current working directory.

@radeksimko I thought I understood this comment:

> How about we create the `product.Product{}` on-the-fly? 

But then when I went to look at it I don't think I understand what you're asking for here.

radeksimko reviewed (commented):
Thanks for the updates @tgross 

This is getting quite close to a merge-able state. I left some suggestions in-line.

kmoe commented on cmd/hc-install/cmd_install.go:
```diff
@@ -0,0 +1,156 @@
+
+  This command installs a HashiCorp product.
+  Options:
+    -version  [REQUIRED] Version of product to install.
```
What if you want to install the latest version (either latest stable (default) or latest prerelease)?

radeksimko replied:
I don't have any too strong opinion on this.

Personally I'd be fine with keeping this required initially and flipping it optional later, which is how one could express _latest stable_? I'm not sure if there's a strong need for _latest pre-release_ in the CLI. If there is, we could probably solve that with another flag - say `-prerelease` which would conflict with `-version` and map directly to https://github.com/hashicorp/hc-install/blob/5bf52541bb816aa9364a71733940c9842e01ba54/releases/latest_version.go#L26

So I think this is solvable later without breaking the API, but if you feel strongly about supporting it now - I'm okay with that as well! 😃

tgross commented on cmd/hc-install/cmd_install.go:
```diff
@@ -0,0 +1,156 @@
+		}
+	default:
+		return fmt.Errorf("project %s cannot be downloaded", project)
+	}
```
The error doesn't distinguish at all between "this product doesn't exist" vs "this version of the product doesn't exist", so I'll make this change and just return the error we get.

tgross commented:
@radeksimko I've updated the PR to resolve all the suggestions except for the question about whether `--version` should be required, which still seems open for discussion? I can squash/rebase this as needed as well too, just let me know.

radeksimko commented on cmd/hc-install/cmd_install.go:
```diff
@@ -0,0 +1,150 @@
+		InstallDir: installDirPath,
+	}
+
+	ctx := context.Background()
```
I was going to originally suggest adding cancellation based on OS signals (like SIGINT) but after re-reading the installation code I realized that we don't actually do much with the context anyway currently, so there's probably not much value in adding it _yet_.

We can take care of it in a separate PR later, where we can also allow graceful cancellation of the underlying HTTP calls and maybe FS operations via context.

kmoe replied:
Opened https://github.com/hashicorp/hc-install/issues/64

radeksimko commented on cmd/hc-install/cmd_install.go:
```diff
@@ -0,0 +1,150 @@
+		c.Ui.Error(msg)
+		return 1
+	}
+
```
```suggestion

c.Ui.Info("installed %s@%s to %s", product, version, installDirPath)
```
Nitpick: Given that we already product informative messages, is it worth also acknowledging the installation succeeded?

_maybe_ even better message could contain full path to the binary, assuming `c.install()` can return it.

tgross replied:
Done.

radeksimko commented on cmd/hc-install/cmd_install.go:
```diff
@@ -0,0 +1,150 @@
+	}
+	product := fs.Args()[0]
+
+	if version == "" {
+		c.Ui.Error("-version flag is required")
+		return 1
+	}
```
As discussed earlier, I'm fine with this behaviour for now. We can still make it optional later and decide later what the right default is and what it means for pre-releases.

@kmoe Do you have any different thoughts here?

kmoe replied:
Yes, I like this approach. Require `-version` for now and we can always make it optional later.

radeksimko reviewed (approved):
Hey @tgross 
Thanks for making these updates.

Aside from my in-line comments, this LGTM.

tgross commented on cmd/hc-install/cmd_install.go:
```diff
@@ -0,0 +1,150 @@
+
+	_, err = io.Copy(dFile, sFile)
+	return err
+}
```
Done

radeksimko commented on cmd/hc-install/cmd_install.go:
```diff
@@ -0,0 +1,122 @@
+	}
+
+	ctx := context.Background()
+	_, err = i.Install(ctx, []src.Installable{source})
+	return filepath.Join(installDirPath, source.Product.BinaryName()), err
```
```suggestion
	return i.Install(ctx, []src.Installable{source})
```
Nitpick: the `Install` method already returns full path, although I admit the interfaces are [not at all documented](https://pkg.go.dev/github.com/hashicorp/hc-install@v0.4.0/src#Installable), so this isn't obvious. 😅 

https://github.com/hashicorp/hc-install/blob/ea027be5cd53c09168f2494a45f3c912ed718e98/releases/exact_version.go#L123-L133

tgross replied:
Fixed.

kmoe commented on cmd/hc-install/cmd_install.go:
```diff
@@ -0,0 +1,120 @@
+	}
+
+	if installDirPath == "" {
+		cwd, err := os.Getwd()
+		if err != nil {
+			c.Ui.Error(fmt.Sprintf("Could not get current working directory for default installation path: %v", err))
+		}
+		installDirPath = cwd
+		return 1
```
The `return 1` needs to be inside the `if`, otherwise the CLI always exits 1 when `-path` is not supplied.

```suggestion
		cwd, err := os.Getwd()
		if err != nil {
			c.Ui.Error(fmt.Sprintf("Could not get current working directory for default installation path: %v", err))
            return 1
		}
		installDirPath = cwd
```
